File: tests/fixtures/python_project/src/services/user_service.py
"""User service with business logic and logging."""

from __future__ import annotations

import logging
from typing import Optional, Protocol

from src.errors.app_error import NotFoundError, ValidationError
from src.models.user import User, UserRole

logger = logging.getLogger(__name__)


class UserRepository(Protocol):
    """Repository protocol for user persistence."""

    def find_by_id(self, user_id: str) -> Optional[User]: ...
    def find_by_email(self, email: str) -> Optional[User]: ...
    def save(self, user: User) -> User: ...
    def delete(self, user_id: str) -> bool: ...


class UserService:
    """Handles user-related business operations."""

    def __init__(self, repository: UserRepository) -> None:
        self._repository = repository

    def get_user(self, user_id: str) -> User:
        """Retrieve a user by ID."""
        logger.info("Fetching user", extra={"user_id": user_id})
        user = self._repository.find_by_id(user_id)
        if user is None:
            logger.warning("User not found", extra={"user_id": user_id})
            raise NotFoundError("User", user_id)
        return user

    def create_user(
        self, username: str, email: str, role: UserRole = UserRole.VIEWER
    ) -> User:
        """Create a new user with validation."""
        logger.info("Creating user", extra={"username": username, "email": email})

        existing = self._repository.find_by_email(email)
        if existing is not None:
            raise ValidationError(f"Email already in use: {email}", field="email")

        user = User(username=username, email=email, role=role)
        saved = self._repository.save(user)
        logger.info("User created", extra={"user_id": saved.id, "role": role.value})
        return saved

    def update_role(self, user_id: str, new_role: UserRole) -> User:
        """Update a user's role."""
        logger.info(
            "Updating user role",
            extra={"user_id": user_id, "new_role": new_role.value},
        )
        user = self.get_user(user_id)
        updated = user.model_copy(update={"role": new_role})
        return self._repository.save(updated)

    def deactivate_user(self, user_id: str) -> User:
        """Deactivate a user account."""
        logger.warning("Deactivating user", extra={"user_id": user_id})
        user = self.get_user(user_id)
        from src.models.user import UserStatus

        updated = user.model_copy(update={"status": UserStatus.INACTIVE})
        return self._repository.save(updated)

File: tests/fixtures/typescript_project/src/services/user.service.ts
/**
 * User service — business logic for user operations.
 * Demonstrates: ESM imports, type-only imports, class-based service, async patterns.
 */

import type { User, CreateUserInput, UpdateUserInput, UserId } from "../types/index.js";

import { UserModel } from "../models/index.js";
import { NotFoundError, ConflictError } from "../errors/index.js";
import { logger } from "../utils/logger.js";

export class UserService {
  private readonly users: Map<string, UserModel> = new Map();

  public async createUser(input: CreateUserInput): Promise<User> {
    logger.info(`Creating user: ${input.email}`);

    const existing = this.findByEmail(input.email);
    if (existing) {
      throw new ConflictError(`User with email '${input.email}' already exists`);
    }

    const user = UserModel.create(input);
    this.users.set(user.id, user);

    logger.info(`User created: ${user.id}`);
    return user.toJSON() as unknown as User;
  }

  public async getUserById(id: UserId): Promise<User> {
    const user = this.users.get(id);
    if (!user) {
      throw new NotFoundError("User", id);
    }
    return user.toJSON() as unknown as User;
  }

  public async updateUser(id: UserId, input: UpdateUserInput): Promise<User> {
    const user = this.users.get(id);
    if (!user) {
      throw new NotFoundError("User", id);
    }

    if (input.name !== undefined) {
      user.updateName(input.name);
    }
    if (input.email !== undefined) {
      const emailConflict = this.findByEmail(input.email);
      if (emailConflict && emailConflict.id !== id) {
        throw new ConflictError(`Email '${input.email}' is already in use`);
      }
      user.updateEmail(input.email);
    }
    if (input.role !== undefined) {
      user.updateRole(input.role);
    }

    logger.info(`User updated: ${id}`);
    return user.toJSON() as unknown as User;
  }

  public async deleteUser(id: UserId): Promise<void> {
    const user = this.users.get(id);
    if (!user) {
      throw new NotFoundError("User", id);
    }

    this.users.delete(id);
    logger.info(`User deleted: ${id}`);
  }

  public async listUsers(page: number = 1, pageSize: number = 20): Promise<{
    items: User[];
    total: number;
  }> {
    const allUsers = Array.from(this.users.values());
    const start = (page - 1) * pageSize;
    const items = allUsers
      .slice(start, start + pageSize)
      .map((u) => u.toJSON() as unknown as User);

    return {
      items,
      total: allUsers.length,
    };
  }

  private findByEmail(email: string): UserModel | undefined {
    for (const user of this.users.values()) {
      if (user.email === email) {
        return user;
      }
    }
    return undefined;
  }
}
